### PR TITLE
OCPBUGS-29110: Reconcile over ICSP/IDMS

### PR DESCRIPTION
**Pull request by jparrill**
During the normal reconciliation loop, this patch allow the Hypershift Operator to grab the new ICSP/IDMS and add them to the CPO, allowing the HCP pods to use the right internal registry image. The changes on the ICSP/IDMS will make the control-plane-operator pod to be restarted in order to update the OPENSHIFT_IMG_OVERRIDES env variable.

**Which issue(s) this PR fixes**:
Fixes #[OCPBUGS-29110](https://issues.redhat.com/browse/OCPBUGS-29110)

- [X] Tested in BareMetal Disconnected environment 

Signed-off-by: Juan Manuel Parrilla Madrid <jparrill@redhat.com>

### CODE REVIEW

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 1]
```diff
@@ -219,6 +219,8 @@ func (r *HostedClusterReconciler) SetupWithManager(mgr ctrl.Manager, createOrUpd
 	// When SCC is not available (Kubernetes), we want to explicitly set a default (non-root) security context
 	r.SetDefaultSecurityContext = !r.ManagementClusterCapabilities.Has(capabilities.CapabilitySecurityContextConstraint)
 
+	r.ReconcileMetadataProviders = r.ReconcileMetadataProvidersImpl
```
Comment (Patryk-Stefanski) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:222]: why are we doing this?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 2]
```diff
@@ -335,6 +337,13 @@ func (r *HostedClusterReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 	return res, err
 }
 
+func (r *HostedClusterReconciler) ReconcileMetadataProvidersImpl(ctx context.Context, imgOverrides map[string]string) (releaseinfo.ProviderWithOpenShiftImageRegistryOverrides, hyperutil.ImageMetadataProvider, error) {
+	var err error
```
Comment (Patryk-Stefanski) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:341]: do we need this var here ?

Comment (jparrill) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:341]: Yep, you're right. Let me change it

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller_test.go [hunk 3]
```diff
@@ -144,7 +146,9 @@ func TestHasBeenAvailable(t *testing.T) {
 				CertRotationScale:             24 * time.Hour,
 				createOrUpdate:                func(reconcile.Request) upsert.CreateOrUpdateFN { return ctrl.CreateOrUpdate },
 				ManagementClusterCapabilities: &fakecapabilities.FakeSupportNoCapabilities{},
-				now:                           func() metav1.Time { return reconcilerNow },
+				ReconcileMetadataProviders: func(ctx context.Context, imgOverrides map[string]string) (releaseinfo.ProviderWithOpenShiftImageRegistryOverrides, hyperutil.ImageMetadataProvider, error) {
+					return &fakereleaseprovider.FakeReleaseProvider{}, &fakeimagemetadataprovider.FakeImageMetadataProvider{Result: &dockerv1client.DockerImageConfig{}}, nil
+				}, now: func() metav1.Time { return reconcilerNow },
```
Comment (Patryk-Stefanski) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller_test.go:151]: nit can we move `now:` to a new line to match below

File: support/globalconfig/imagecontentsource_test.go [hunk 4]
```diff
@@ -173,3 +178,155 @@ func createFakeIDMS() *configv1.ImageDigestMirrorSetList {
 		},
 	}
 }
+func TestRenconcileMgmtImageRegistryOverrides(t *testing.T) {
```
Comment (Patryk-Stefanski) [inline support/globalconfig/imagecontentsource_test.go:181]: typo
```suggestion
func TestReconcileMgmtImageRegistryOverrides(t *testing.T) {
```

**Pull request merged**